src/Microsoft.IdentityModel.JsonWebTokens/Json/JsonClaimSet.cs

```diff
@@ -18,121 +19,134 @@ namespace Microsoft.IdentityModel.JsonWebTokens
     /// </summary>
     internal class JsonClaimSet
     {
-        internal static JsonClaimSet Empty { get; } = new JsonClaimSet("{}"u8.ToArray());
+        internal const string ClassName = "Microsoft.IdentityModel.JsonWebTokens.JsonClaimSet";
 
+        internal static JsonClaimSet Empty { get; } = new JsonClaimSet("{}"u8.ToArray());
+        internal object _claimsLock = new();
+        internal IDictionary<string, object> _jsonClaims;
         private IList<Claim> _claims;
 
-        internal JsonClaimSet(JsonDocument jsonDocument)
+        internal JsonClaimSet(IDictionary<string, object> jsonClaims)
         {
-            // This method is assuming ownership of the JsonDocument, which is backed by one or more ArrayPool arrays.
-            // We need to dispose of it to avoid leaking arrays from the pool.  To achieve that, we clone the root element,
-            // which will result in a new JsonElement being created that's not tied to the original and that's not backed by
-            // ArrayPool memory, after which point we can dispose of the original to return the array(s) to the pool.
-            RootElement = jsonDocument.RootElement.Clone();
-            jsonDocument.Dispose();
+            _jsonClaims = jsonClaims;
         }
-
-        internal JsonClaimSet(byte[] jsonBytes) : this(JsonDocument.Parse(jsonBytes))
+        internal JsonClaimSet(byte[] jsonUtf8Bytes)
         {
+            _jsonClaims = JwtTokenUtilities.CreateClaimsDictionary(jsonUtf8Bytes, jsonUtf8Bytes.Length);
         }
 
-        internal JsonClaimSet(string json) : this(JsonDocument.Parse(json))
+        internal JsonClaimSet(string json) : this(Encoding.UTF8.GetBytes(json))
```

Comment:
Is this one of the places we could use the ArrayPool so we don't need to allocate a new `byte[]` just to turn it into a Dictionary and throw the byte array away?